File: TripKitAndroidUI/src/main/java/com/skedgo/tripkit/ui/model/DeparturesResponse.java
package com.skedgo.tripkit.ui.model;

import androidx.annotation.Nullable;
import com.google.gson.annotations.SerializedName;
import com.skedgo.tripkit.common.model.RealtimeAlert;
import com.skedgo.tripkit.common.model.ScheduledStop;

import java.util.ArrayList;
import java.util.List;

/**
 * @see <a href="https://redmine.buzzhives.com/projects/buzzhives/wiki/RealTime_API#DepartureServlet-new-servlet-departuresjson">departures.json API</a>
 */
public class DeparturesResponse {
  /**
   * (Optional)
   */
  @SerializedName("error")
  public String error;

  /**
   * (Optional)
   */
  @SerializedName("usererror")
  public boolean hasError;

  @SerializedName("embarkationStops")
  public List<ServicesResponse> embarkationStopList;

  /**
   * (Optional)
   */
  @SerializedName("stops")
  public List<ScheduledStop> stopList;

  /**
   * (Optional)
   */
  @SerializedName("parentInfo")
  private ScheduledStop parentInfo;

  /**
   * (Optional)
   */
  @SerializedName("alerts")
  private List<RealtimeAlert> alerts;

  private List<TimetableEntry> mServiceList;

  @Nullable
  public ScheduledStop getParentInfo() {
    return parentInfo;
  }

  /**
   * Assigns stop code into corresponding service
   * <p/>
   * NOTE: Must call this method after parsing the response
   */
  public void processEmbarkationStopList() {
    if (embarkationStopList != null) {
      for (DeparturesResponse.ServicesResponse servicesResponse : embarkationStopList) {
        if (servicesResponse.serviceList != null) {
          for (TimetableEntry service : servicesResponse.serviceList) {
            if (service != null) {
              service.setStopCode(servicesResponse.stopCode);
            }
          }
        }
      }
    }
  }

  @Nullable
  public List<TimetableEntry> getServiceList() {
    if (mServiceList == null) {
      mServiceList = extractServiceList();
    }

    return mServiceList;
  }

  @Nullable
  public List<RealtimeAlert> getAlerts() {
    return alerts;
  }

  public void setAlerts(List<RealtimeAlert> alerts) {
    this.alerts = alerts;
  }

  /**
   * Extracts services from the embarkation stops
   */
  @Nullable
  private List<TimetableEntry> extractServiceList() {
    if (embarkationStopList == null) {
      return null;
    }

    List<TimetableEntry> serviceList = new ArrayList<TimetableEntry>();
    for (ServicesResponse servicesResponse : embarkationStopList) {
      if (servicesResponse.serviceList != null) {
        serviceList.addAll(servicesResponse.serviceList);
      }
    }

    return serviceList;
  }

  public static class ServicesResponse {
    @SerializedName("stopCode")
    public String stopCode;

    @SerializedName("services")
    public List<TimetableEntry> serviceList;
  }
}

File: TripKitAndroidUI/src/main/java/com/skedgo/tripkit/ui/model/DeparturesResponse.kt
package com.skedgo.tripkit.ui.model

import com.google.gson.annotations.SerializedName
import com.skedgo.tripkit.common.model.realtimealert.RealtimeAlert
import com.skedgo.tripkit.common.model.stop.ScheduledStop

/**
 * @see [departures.json API](https://redmine.buzzhives.com/projects/buzzhives/wiki/RealTime_API.DepartureServlet-new-servlet-departuresjson)
 */
class DeparturesResponse {
    /**
     * (Optional)
     */
    @SerializedName("error")
    var error: String? = null

    /**
     * (Optional)
     */
    @SerializedName("usererror")
    var hasError: Boolean = false

    @SerializedName("embarkationStops")
    var embarkationStopList: List<ServicesResponse>? = null

    /**
     * (Optional)
     */
    @SerializedName("stops")
    var stopList: List<ScheduledStop>? = null

    /**
     * (Optional)
     */
    @SerializedName("parentInfo")
    val parentInfo: ScheduledStop? = null

    /**
     * (Optional)
     */
    @SerializedName("alerts")
    var alerts: List<RealtimeAlert>? = null

    private var mServiceList: List<TimetableEntry>? = null

    /**
     * Assigns stop code into corresponding service
     *
     *
     * NOTE: Must call this method after parsing the response
     */
    fun processEmbarkationStopList() {
        embarkationStopList?.forEach { servicesResponse ->
            servicesResponse.serviceList?.forEach { service ->
                service.stopCode = servicesResponse.stopCode
            }
        }
    }

    val serviceList: List<TimetableEntry>?
        get() {
            if (mServiceList == null) {
                mServiceList = extractServiceList()
            }

            return mServiceList
        }

    /**
     * Extracts services from the embarkation stops
     */
    private fun extractServiceList(): List<TimetableEntry>? {
        if (embarkationStopList == null) {
            return null
        }

        val serviceList: MutableList<TimetableEntry> = ArrayList()
        for (servicesResponse in embarkationStopList!!) {
            if (servicesResponse.serviceList != null) {
                serviceList.addAll(servicesResponse.serviceList!!)
            }
        }

        return serviceList
    }

    data class ServicesResponse(
        @SerializedName("stopCode") val stopCode: String,
        @SerializedName("services") val serviceList: List<TimetableEntry>?
    )
}
